File: macos/Sources/NoctusCore/ViewModels/MainViewModel.swift
import Foundation

public final class MainViewModel {
    private let fileOps: FileOperationsProtocol
    private let settings: SettingsStore

    public let leftPane: PaneViewModel
    public let rightPane: PaneViewModel
    public let commandRegistry: CommandRegistry
    public let keyBindingResolver: KeyBindingResolver
    public let bookmarkStore: BookmarkStore
    public let dropStack: DropStackService

    public var splitMode: SplitMode = .vertical
    public var splitRatio: Double = 0.5

    public var activePane: PaneViewModel { leftPane.isActive ? leftPane : rightPane }
    public var inactivePane: PaneViewModel { leftPane.isActive ? rightPane : leftPane }

    public init(
        shellService: ShellServiceProtocol,
        fileOps: FileOperationsProtocol,
        fileWatcher: FileWatcherProtocol,
        settings: SettingsStore,
        commandRegistry: CommandRegistry,
        keyBindingResolver: KeyBindingResolver,
        bookmarkStore: BookmarkStore,
        dropStack: DropStackService
    ) {
        self.fileOps = fileOps
        self.settings = settings
        self.commandRegistry = commandRegistry
        self.keyBindingResolver = keyBindingResolver
        self.bookmarkStore = bookmarkStore
        self.dropStack = dropStack

        self.leftPane = PaneViewModel(shellService: shellService, fileWatcher: fileWatcher)
        self.rightPane = PaneViewModel(shellService: shellService, fileWatcher: fileWatcher)
        leftPane.isActive = true
    }

    public func switchActivePane() {
        leftPane.isActive = !leftPane.isActive
        rightPane.isActive = !rightPane.isActive
    }

    public func toggleSplitMode() {
        switch splitMode {
        case .vertical: splitMode = .horizontal
        case .horizontal: splitMode = .single
        case .single: splitMode = .vertical
        }
    }

    public func copyToOtherPane() async throws {
        guard let selection = activePane.activeTab?.selection,
              !selection.isEmpty,
              let destination = inactivePane.activeTab?.location else { return }
        let sources = selection.map(\.path)
        try await fileOps.copy(sources, to: destination)
    }

    public func moveToOtherPane() async throws {
        guard let selection = activePane.activeTab?.selection,
              !selection.isEmpty,
              let destination = inactivePane.activeTab?.location else { return }
        let sources = selection.map(\.path)
        try await fileOps.move(sources, to: destination)
    }

    public func syncNavigation() async throws {
        guard let activeLocation = activePane.activeTab?.location,
              inactivePane.activeTab != nil else { return }
        try await inactivePane.activeTab?.navigate(to: activeLocation)
    }

    public func saveSession() {
        settings.set("session.splitMode", value: splitMode.rawValue)
        settings.set("session.splitRatio", value: splitRatio)
    }

    public func restoreSession() {
        let modeStr: String = settings.get("session.splitMode", default: "vertical")
        if let mode = SplitMode(rawValue: modeStr) {
            splitMode = mode
        }
        splitRatio = settings.get("session.splitRatio", default: 0.5)
    }
}

File: src/Explorer.Core/ViewModels/MainViewModel.cs
using NoctusExplorer.Core.Abstractions;
using NoctusExplorer.Core.Models;
using NoctusExplorer.Core.Services;
using ReactiveUI;

namespace NoctusExplorer.Core.ViewModels;

public class MainViewModel : ReactiveObject
{
    private readonly IFileOperations _fileOps;
    private readonly SettingsStore _settings;
    private SplitMode _splitMode = SplitMode.Vertical;
    private double _splitRatio = 0.5;

    public MainViewModel(
        IShellService shellService,
        IFileOperations fileOps,
        IFileWatcher fileWatcher,
        SettingsStore settings,
        CommandRegistry commandRegistry,
        KeyBindingResolver keyBindingResolver,
        BookmarkStore bookmarkStore,
        CustomActionStore customActionStore,
        DropStackService dropStack,
        OperationsQueue operationsQueue)
    {
        _fileOps = fileOps;
        _settings = settings;

        LeftPane = new PaneViewModel(shellService, fileWatcher);
        RightPane = new PaneViewModel(shellService, fileWatcher);

        CommandRegistry = commandRegistry;
        KeyBindingResolver = keyBindingResolver;
        BookmarkStore = bookmarkStore;
        CustomActionStore = customActionStore;
        DropStack = dropStack;
        OperationsQueue = operationsQueue;

        // Left pane starts active
        LeftPane.IsActive = true;
    }

    public PaneViewModel LeftPane { get; }
    public PaneViewModel RightPane { get; }

    public PaneViewModel ActivePane => LeftPane.IsActive ? LeftPane : RightPane;
    public PaneViewModel InactivePane => LeftPane.IsActive ? RightPane : LeftPane;

    public SplitMode SplitMode
    {
        get => _splitMode;
        set => this.RaiseAndSetIfChanged(ref _splitMode, value);
    }

    public double SplitRatio
    {
        get => _splitRatio;
        set => this.RaiseAndSetIfChanged(ref _splitRatio, value);
    }

    public CommandRegistry CommandRegistry { get; }
    public KeyBindingResolver KeyBindingResolver { get; }
    public BookmarkStore BookmarkStore { get; }
    public CustomActionStore CustomActionStore { get; }
    public DropStackService DropStack { get; }
    public OperationsQueue OperationsQueue { get; }

    public void SwitchActivePane()
    {
        LeftPane.IsActive = !LeftPane.IsActive;
        RightPane.IsActive = !RightPane.IsActive;
        this.RaisePropertyChanged(nameof(ActivePane));
        this.RaisePropertyChanged(nameof(InactivePane));
    }

    public void ToggleSplitMode()
    {
        SplitMode = SplitMode switch
        {
            SplitMode.Single => SplitMode.Vertical,
            SplitMode.Vertical => SplitMode.Horizontal,
            SplitMode.Horizontal => SplitMode.Single,
            _ => SplitMode.Vertical
        };
    }

    public void CopyToOtherPane()
    {
        var selection = ActivePane.ActiveTab?.Selection;
        var destination = InactivePane.ActiveTab?.Location;
        if (selection is null || selection.Count == 0 || destination is null) return;

        CopyToOtherPane(selection.Select(e => e.Path).ToList(), destination);
    }

    public void CopyToOtherPane(IReadOnlyList<PathRef> sources, PathRef destination)
    {
        if (sources.Count == 0) return;
        var handle = _fileOps.Copy(sources, destination);
        OperationsQueue.Enqueue(handle);
    }

    public void MoveToOtherPane()
    {
        var selection = ActivePane.ActiveTab?.Selection;
        var destination = InactivePane.ActiveTab?.Location;
        if (selection is null || selection.Count == 0 || destination is null) return;

        MoveToOtherPane(selection.Select(e => e.Path).ToList(), destination);
    }

    public void MoveToOtherPane(IReadOnlyList<PathRef> sources, PathRef destination)
    {
        if (sources.Count == 0) return;
        var handle = _fileOps.Move(sources, destination);
        OperationsQueue.Enqueue(handle);
    }

    public async Task SyncNavigationAsync(CancellationToken ct = default)
    {
        var activeLocation = ActivePane.ActiveTab?.Location;
        if (activeLocation is null || InactivePane.ActiveTab is null) return;
        await InactivePane.ActiveTab.NavigateAsync(activeLocation, ct);
    }

    public void SaveSession()
    {
        _settings.Set("session.splitMode", SplitMode.ToString());
        _settings.Set("session.splitRatio", SplitRatio);
        // Tab state serialization would be expanded in full implementation
    }

    public void RestoreSession()
    {
        var modeStr = _settings.Get("session.splitMode", "Vertical");
        if (Enum.TryParse<SplitMode>(modeStr, out var mode))
            SplitMode = mode;
        SplitRatio = _settings.Get("session.splitRatio", 0.5);
    }
}
